storage/replica_raftstorage.go

```diff
@@ -520,19 +507,6 @@ func (r *Replica) applySnapshot(batch engine.Engine, snap raftpb.Snapshot) (uint
 		}
```

Comment:
While iterating over `snapData.KV` we should also drop any unreplicated keys. This is an extra layer of safety against a snapshot that includes unreplicated data, so it won't actually do anything once you make the change in my previous comment. But I think this change as it stands is incorrect since the HardState will be overwritten by the one from the snapshot. This suggests a gap in our test coverage, but I'm not sure how exactly you would engineer a situation in which transferring HardState across snapshots would actually make a difference.